### PR TITLE
GdWheel policy + AccessEvent based admission

### PR DESCRIPTION
**Pull request by himelbrand**
I've Implemented GDWheel policy: [pdf](https://conglongli.github.io/paper/gdwheel-eurosys2015.pdf)
This policy is cost aware, where the cost can be either the weight or penalty (In the paper they use cost as the recomputation cost for an item, which is basicly the miss penalty).

In addition I've extended the Admittor so we now have EventsAdmittor, which uses the AccessEvent instead of just using the key, this is usful for size or latency based admission policies.

### CODE REVIEW

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java
Comment (ben-manes) [inline simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java:54]: Please make these cases multiple lines (we casually follow [Google's style guide](https://google.github.io/styleguide/javaguide.html)). For languages like Kotlin / Scala this is common for one-liners, but isn't in Java.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java
Comment (ben-manes) [inline simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java:64]: Maybe we should flip this around? The admittor was written prior to having an AccessEvent, where we introduced `KeyOnly` types for conversion. So we could make `KeyOnlyAdmittor` as a convenience for existing usages, while making the base interface event-aware.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java
Comment (ben-manes) [inline simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: You might want to use `TreeSet` instead, which is implemented as a red-black tree. That has a O(lg n) operation for `remove(Object)`, whereas `PriorityQueue` is a heap with O(n) time.

> Implementation note: this implementation provides O(log(n)) time for the enqueuing and dequeuing methods
offer, poll, remove(), and add; linear time for the remove(Object) and contains(Object) methods; and constant time for the retrieval methods peek, element, and size.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java
Comment (ben-manes) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java:54]: Actually, you can ignore this. I'll go back and fix up minor style nits (line Google's use of 2-space indent) after this is merged.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java
Comment (ben-manes) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java:64]: If you're okay with this proposal, I can go back and migrate towards that. That way it doesn't dirty up your PR and be an isolated refactoring.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java
Comment (ben-manes) [inline simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:217]: fyi, Java 8 introduced a one-liner for these cases by using lambdas, 
`Comparator.comparingDouble(AccessEvent:: missPenalty)`

**Review by ben-manes — APPROVED** [review]
Thanks for implementing this! I'll merge after the comments, and then go back to do some minor tidying up for consistency, etc.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java
Comment (vlsi) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: PriorityQueue + HashMap all the way :)

Comment (ben-manes) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: In this case, `remove(Object)` is called on every cache hit to reorder the entry. Since that is an O(n) operation, the performance degrades as the cache size increases.

`PriorityQueue` is optimal when used a a queue, where specific entries are not commonly searched for. For example Java's `ScheduledExecutorService` uses a min-heap and marks the entry as dead when cancelled, rather then removing it. In that case the cheaper algorithmic cost is beneficial, especially since a scheduler rarely has millions of events. Here the remove + add to resort the entry on every hit could lead to higher execution times and make GDWheel appear slower than it might be in practice.

Comment (vlsi) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: I'm always forgetting that `PriorityQueue` misses `removeAt(int)` method, so adding `HashMap` does not solve the issue. BinaryHeap with node removal is needed.

Tombstones would create an ABA issue, so it might be easier to implement or borrow a binary heap if priorityqueue with removals is needed.

> Here the remove + add to resort the entry on every hit could lead to higher execution times and make GDWheel appear slower than it might be in practice

The paper suggests that Queue is a linked list:

>Since each queue is implemented by a doubly linked list, removing or inserting a given node requires only constant time

In other words, the queue needs:
* Fast append to `head`
* Fast removal from `tail`
* Fast removal from the middle

It is funny how `LinkedHashSet` almost works here. The sad part is `LinkedHashSet` does not expose `reverseIterator` even though it maintains double-linked list.

Comment (ben-manes) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: I'll have to reread the GDWheel paper, as it has been a few years. I believe Apache Commons' `ListOrderedMap` is that more flexible variant that would be simple to adapt here, based on your comment. In our [TimerWheel](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/TimerWheel.java) we use intrusive linked lists for expiration to maintain an O(1) priority queue. However that code is very optimized while this is simulator code for algorithm analysis. Ideally the simulator policies would be clean, readable reference implementations that are easy to adapt for others. They don't tackle messy topics like concurrency, though, and are strictly for analysis for exploring each algorithm's insights.

Comment (vlsi) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: `ListOrderedMap` uses `ArrayList` which implies `O(N)` removals.

What do you think of the following?
The below structure efficiently implements "add to head" (via `put(k, v)`), "remove from tail" (via `removeEldestEntry` trick) and "remove".

```java
public class DequeWithRemoval<K, V> extends LinkedHashMap<K, V> {
    private static final Object SENTINEL = new Object();
    private @Nullable V lastValue;
    private boolean removeLastRequested;

    public DequeWithRemoval(int initialCapacity, float loadFactor) {
        super(initialCapacity, loadFactor, /* accessOrder */ false);
    }

    public V removeLast() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        }
        removeLastRequested = true;
        // put triggers removeEldestEntry which is how we retrieve the latest element
        //noinspection unchecked
        ((Map) this).put(SENTINEL, SENTINEL);
        V lastValue = this.lastValue;
        this.lastValue = null;
        // Remove sentinel
        values().iterator().remove();

        return lastValue;
    }

    @Override
    protected boolean removeEldestEntry(Map.Entry<K, V> eldest) {
        if (!removeLastRequested) {
            return false;
        }
        lastValue = eldest.getValue();
        return true;
    }
}
```

Comment (ben-manes) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: oh man, I didn't realize it was done so poorly in Apache. I suppose it's my fault for never using it an assuming it was on par if I ever found a need.

That's neat. I don't mind implementing intrusive linked lists. If I was to write implement this policy, I would include the prev/next references on the `Node` and handle it directly. This is done in many of the other policies for LRU ordering. The code is very simple and straightforward.

Comment (ben-manes) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: Oh, it was [LinkedMap](https://github.com/rkanter/collections-generic/blob/master/src/main/java/org/apache/commons/collections15/map/LinkedMap.java) that was like `LinkedHashMap`.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java
Comment (himelbrand) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/Admittor.java:64]: Sounds good, I'm good with your proposal.
Thanks

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java
Comment (himelbrand) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:166]: I will change it to TreeSet, since according to the paper it should be O(log(n)) in the worst case and amortized is O(1), so O(n) for a single operation is not ideal for the implemantation.
Thanks for the info.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java
Comment (himelbrand) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:217]: Cool, I like this syntax better, I'll adopt it.

**Comment by himelbrand** [comment]
I've just updated the few things you mentioned @ben-manes 
I pushed the changes now

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java
Comment (vlsi) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java]: @himelbrand , could you please clarify where do you get `according to the paper it should be O(log(n))` from?

https://conglongli.github.io/paper/gdwheel-eurosys2015.pdf

> 3.2.2 GD-Wheel Time Complexity Analysis
>Since each
>queue is implemented by a doubly linked list, removing
>or inserting a given node requires only constant time. An
>insertion of a given object requires one list insertion. A
>reuse of a given object requires one list removal and one list
>insertion. Thus handling either an insertion or a reuse of an
>object takes only O(1) time in the worst case

I do not see a priority queue here. However, I must admit it is the first time I see the paper, so I can be wrong.

Comment (himelbrand) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java]: @vlsi, the are 2 published versions of this paper, but I think this appears in both of them.
specifically regarding the priority queue, it is definitely mentioned in the pseudo code of the algorithm and a few extra locations.

regarding the time complexity, if you keep going a bit further in 3.2.2 you'll see that O(1) is amortized. I might have been confused, as I now see that they mentioned O(n), but I might remember the log(n) from the other version when compared to the gree dual approach.

Comment (vlsi) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java]: PriorityQueue is mentioned in `Algorithm 1 The GreedyDual Algorithm under Cao et al.’s implementation`
However, point of adding `GD-Wheel` policy is to replace priority queues with multiple linked lists.

O(n) is related to `migration` which indeed scans the queue.

>paper: As a result, **objects with different priorities** are stored in **different queues**

So I'm inlined that the queues should be implemented as linked lists so the implementation matches the paper.

Comment (himelbrand) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java]: When I implemented it I did it according to [pdf](https://www.cs.cmu.edu/~conglonl/gdwheel-ladis2013.pdf).
which is the Algorithm 1 The GreedyDual Algorithm under Cao et al.’s implementation.
the paper I gave the link and that you cited, was not the one I was reading when I implemented the algorithm.
This is the reason for the confusion.

It could be nice to implement the other version too, or to add the possibility to switch between the 2 variants in the future.

Comment (ben-manes) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java]: Let me try to spend some time reading the two papers and seeing if I can optimize it while matching the hit rates. I was planning on refactoring the code a bit to pull apart the long `record` method, e.g. to match the other policies which use `onHit`, `onMiss`, `evict`, etc. I'd also like to extract paper comments to help guide the reader, e.g. as done in `GdfsPolicy`. This will all take a bit longer, but the code is a good baseline to refactor on top of.

**Comment by ben-manes** [comment]
@himelbrand I quick sanity test and I think the behavior is broken somehow. I was just doing minor formatting changes prior to reading the paper(s), and added a validation check which failed. For the default lirs trace, multi1, it reports a map size of `2606` whereas the maximum is `512`. I reproduced this with your commit and w/o my formatting changes.

```java
@Override
public void finished() {
  checkState(data.size() <= maximumSize, "%s > %s", data.size(), maximumSize);
}
```

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java
Comment (ben-manes) [inline simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:193]: This is the bug that causes the cache to not evict correctly. The method`first()` peeks into the queue, whereas you want `pollFirst()` to remove it.

Comment (ben-manes) [reply simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java:193]: This fix only works if always allowing admission, as if `admit = false` then it seems to break...

**Comment by ben-manes** [comment]
I rewrote the implementation onto O(1) intrusive lists as described by the paper, where I used a sentinel node for simplicity. The algorithm's steps are annotated from the pseudo code to help future readers follow along. For an unknown reason, I did observe a large hit rate improvement on `mutli1.trace` when comparing against the original version, and it now more closely matches GdfsPolicy's hit rate. I removed the usage of an admission policy since I don't think you were actually using it and that caused weird failures in your version.

This was a bit more invasive of a refactor than I had expected, so I apologize for that. I think it more closely matches the paper and is clearer to follow. It helped a lot to have yours and the paper as a reference.

**Comment by vlsi** [comment]
Just in case, it turns out `LinkedHashMap` is good enough (see the updated https://github.com/ben-manes/caffeine/pull/529#discussion_r611101853)
`removeLast` is a mere `iterator().next().remove()`. `removeEldestEntry` is not really needed.

**Pull request closed without merging**
